File: javascript/67.add-binary.js
/*
 * @lc app=leetcode id=67 lang=javascript
 *
 * [67] Add Binary
 */

// @lc code=start
/**
 * @param {string} a
 * @param {string} b
 * @return {string}
 */
var addBinary = function (a, b) {
  // Solution 1:
  // const a_digits = a.split(""),
  //   b_digits = b.split("");
  // let carry = 0;
  // const longerLength = Math.max(a_digits.length, b_digits.length);
  // let result = "";
  // for (let i = 0; i < longerLength; i++) {
  //   const ad = +a_digits.pop() || 0,
  //     bd = +b_digits.pop() || 0;
  //   let s = ad + bd + carry;
  //   if (s >= 2) {
  //     s -= 2;
  //     carry = 1;
  //   } else carry = 0;
  //   result = s + result;
  // }
  // if (carry) result = carry + result;
  // return result;

  // Solution 2:
  return (BigInt(`0b${a}`) + BigInt(`0b${b}`)).toString(2);
};
// @lc code=end

File: rust/67.add-binary.rs
/*
 * @lc app=leetcode id=67 lang=rust
 *
 * [67] Add Binary
 */

// @lc code=start
impl Solution {
    pub fn add_binary(a: String, b: String) -> String {
        let mut result = String::new();
        let mut ac = a.chars().rev();
        let mut bc = b.chars().rev();
        let mut s = 0;
        loop {
            match (ac.next(), bc.next()) {
                (Some(c), Some(d)) => {
                    s = c.to_digit(2).unwrap() + d.to_digit(2).unwrap() + s;
                }
                (Some(c), None) | (None, Some(c)) => {
                    s = c.to_digit(2).unwrap() + s;
                }
                _ => break,
            }
            result.insert(0, char::from_digit(s % 2, 2).unwrap());
            s = (s >= 2) as u32;
        }
        if s != 0 {
            result.insert(0, char::from_digit(1, 2).unwrap());
        }
        return result;
    }
}
// @lc code=end
